### PR TITLE
feat: [84] Create AmountParser utility for smart amount input

### LINKED ISSUE #84: Create AmountParser utility for smart amount input
### Overview
Create a utility that parses the Honey-style smart input format where users type a description with an embedded math expression, and text in parentheses is ignored.

### Examples
| Input | Amount (cents) | Description |
|-------|---------------|-------------|
| `"4*15 zoo tickets (100 in parentheses is ignored)"` | 6000 | `"zoo tickets"` |
| `"25.50"` | 2550 | `""` |
| `"100+50 groceries"` | 15000 | `"groceries"` |
| `"42"` | 4200 | `""` |
| `"rent"` | 0 | `"rent"` |
| `""` | 0 | `""` |

### Implementation

**`app/Support/AmountParser.php`**
- Static method: `parse(string $input): AmountParseResult`
- Algorithm:
  1. Strip all text inside parentheses: `preg_replace('/\([^)]*\)/', '', $input)`
  2. Extract leading math expression (numbers, `*`, `+`, `-`, `/`, `.`, spaces)
  3. Evaluate safely without `eval` — split on `+`/`-` for terms, then `*`/`/` for factors
  4. Convert to cents: `(int) round($result * 100)`
  5. Remaining text after the math expression becomes the description

**`app/Support/AmountParseResult.php`**
- `final readonly class` with `public int $amount` (cents) and `public string $description`

### Tests
- `tests/Unit/Support/AmountParserTest.php`
- Dataset-driven with all examples above plus edge cases:
  - No number input (description only)
  - Empty input
  - Only parenthetical text
  - Decimal amounts
  - Multiple math operators
  - Negative results
  - Very large numbers

### Acceptance Criteria
- [ ] Parser handles all documented input formats
- [ ] No use of `eval()` — safe math evaluation only
- [ ] DTO is readonly with typed properties
- [ ] Dataset-driven unit tests pass
- [ ] `op ci` passes

### Dependencies
None.

---
Part of the **Add/Edit Transaction Modal** feature.

### PR DESCRIPTION
**Pull request by robwilde**
## Summary
- Add `AmountParser::parse()` utility that extracts a computed amount (in cents) and description from Honey-style smart input (e.g. `"4*15 zoo tickets"` → 6000 cents, `"zoo tickets"`)
- Safe math evaluation with operator precedence (`*`/`/` before `+`/`-`) — no `eval()`
- Parenthesized text is stripped and ignored (e.g. `"(notes here)"`)

## New Files
- `app/Support/AmountParseResult.php` — `final readonly` value object with `int $amount` (cents) and `string $description`
- `app/Support/AmountParser.php` — static parser with two-pass tokenizer for safe math evaluation
- `tests/Unit/Support/AmountParserTest.php` — 17 dataset-driven unit tests covering all documented formats and edge cases

## Test plan
- [x] All 15 dataset cases pass (math expressions, decimals, description-only, empty, parenthetical, division by zero, etc.)
- [x] Operator precedence verified (`10+5*2` = 2000 cents, not 3000)
- [x] `op ci` passes clean (lint, PHPStan 0 errors, 661 tests)

Closes #84

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/Support/AmountParser.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    /** @return list<float|string> */
+    private static function tokenize(string $expression): array
+    {
+        $cleaned = preg_replace('/\s+/', '', $expression);
+        preg_match_all('/(\d+\.?\d*|[+\-*\/])/', (string) $cleaned, $matches);
+
+        $tokens = [];
+
+        foreach ($matches[0] as $token) {
+            $tokens[] = is_numeric($token) ? (float) $token : $token;
```
[the remark this answers is not in the document]
Comment (robwilde) [reply app/Support/AmountParser.php:62]: Fixed. Tokenizer regex updated from `(\d+\.?\d*)` to `(\d*\.?\d+)` — now accepts leading-dot decimals like `.50`. Test added: `'leading-dot decimal' => ['.50 tip', 50, 'tip']`.

File: app/Support/AmountParser.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        return $tokens;
+    }
```
[the remark this answers is not in the document]
Comment (robwilde) [reply app/Support/AmountParser.php:66]: Fixed. Tokenizer now prepends `0` when the expression starts with `+` or `-`, so `-10` becomes `0-10` which the two-pass evaluator handles correctly. Tests added for both `-10 refund` and `+25 bonus`.

File: app/Support/AmountParser.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+declare(strict_types=1);
+
+namespace App\Support;
+
+final class AmountParser
+{
+    public static function parse(string $input): AmountParseResult
+    {
+        $stripped = preg_replace('/\([^)]*\)/', '', $input);
```
[the remark this answers is not in the document]
Comment (robwilde) [reply app/Support/AmountParser.php]: Fixed using the suggested iterative approach — `while` loop strips innermost parentheses first until none remain. Handles nested cases like `(April (late))` correctly. Test added: `'nested parentheses' => ['50 rent (April (late))', 5000, 'rent']`.

File: app/Support/AmountParser.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+     * @return list<float|string>
+     */
+    private static function resolveMultiplicationAndDivision(array $tokens): array
+    {
+        $result = [$tokens[0]];
+
+        for ($i = 1, $iMax = count($tokens); $i < $iMax; $i += 2) {
+            $operator = $tokens[$i];
+            $right = $tokens[$i + 1] ?? 0.0;
+
```
[the remark this answers is not in the document]
Comment (robwilde) [reply app/Support/AmountParser.php:79]: Acknowledged but skipping for now — the leading-expression regex already constrains what reaches the evaluator, and malformed sequences like `10+` or `10++5` are unlikely from user input. The fallback to `0.0` for missing tokens is sufficient. Will revisit if real-world usage surfaces edge cases.

File: tests/Unit/Support/AmountParserTest.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+test('parse returns AmountParseResult instance', function () {
+    expect(AmountParser::parse('42'))->toBeInstanceOf(AmountParseResult::class);
+});
+
+test('AmountParseResult is readonly with typed properties', function () {
+    $result = new AmountParseResult(4200, 'groceries');
+
+    expect($result->amount)->toBe(4200)
+        ->and($result->description)->toBe('groceries');
```
[the remark this answers is not in the document]
Comment (robwilde) [reply tests/Unit/Support/AmountParserTest.php:18]: Fixed — renamed test to `'AmountParseResult has typed properties'`. Skipped the readonly write assertion since `final readonly class` enforces immutability at the language level — testing PHP's own guarantees isn't valuable here.
